internal/backend/runtime/omni/controllers/omni/machine_extensions.go

```diff
@@ -213,9 +232,88 @@ func (ctrl *MachineExtensionsController) Reconcile(ctx context.Context,
 	return tracker.cleanup(ctx)
 }
 
-func (ctrl *MachineExtensionsController) getRelatedClusterMachines(ctx context.Context,
-	r controller.QRuntime, configuration *omni.ExtensionsConfiguration,
-) ([]*omni.ClusterMachine, error) {
+// shouldRecalculateExtensions is a workaround built to preserve the existing wrong order of extensions to prevent machines from being upgraded unexpectedly.
+//
+// It does it by a clever trick:
+//
+// - If the version is zero, it means this is a create operation - calculate the extensions again.
+// - If the helpers.InputResourceVersionAnnotation were set AND they have changed on the update, calculate the extensions again.
+// - If the input versions are not yet set and if this is an update operation, we want to keep the existing extensions, even if they were calculated incorrectly.
+func (ctrl *MachineExtensionsController) shouldRecalculateExtensions(me *omni.MachineExtensions, configs []*omni.ExtensionsConfiguration) bool {
+	_, versionsWereSet := me.Metadata().Annotations().Get(helpers.InputResourceVersionAnnotation)
+	versionsAreUpdated := helpers.UpdateInputsVersions(me, configs...)
+
+	isCreate := me.Metadata().Version().Value() == 0
+	if isCreate {
+		return true
+	}
+
+	if versionsWereSet && versionsAreUpdated {
+		return true
+	}
+
+	return false
+}
```

Comment:
@Unix4ever @oguzkilcan @smira this is the idea to avoid unwanted upgrades, can you pls have a look? it also has a test.